### DISCUSSION TITLE
GDB thread awareness info running on QEMU target

### DISCUSSION THREAD

**Discussion by rflmota** [zephyrproject-rtos/zephyr#57816] [Q&A]
Hello, I'm current learning _Zephyr_ and was trying to run some samples on the provided qemu targets (`qemu_cortex_m3`), although I noticed when attaching with `gdb` there's no thread information provided by `threads` command. Is it possible to get thread awareness information with gdb running on qemu targets?

**Comment by cfriedt** [comment]
Hi @rflmota - It's not super obvious, but what I often do when I debug is 
```shell
west build -p auto -b board -t debugserver path/to/app \
    -- \
    -DCONFIG_NO_OPTIMIZATIONS=y \
    -DCONFIG_DEBUG=y \
    -DCONFIG_DEBUG_INFO=y \
    -DCONFIG_DEBUG_THREAD_INFO=y
```

Those extra options have the following effects:
1. "Optimize nothing" (prevent symbols from being "optimized-out")
2. "Build kernel with debugging enabled"
3. "System debugging information" (additional info for debuggers)
4. "Thread awareness support" (allow gdb, openocd, etc to see thread info)

You can always navigate Kconfig options through `west build -t menuconfig` sort of like it's done with the Linux kernel or busybox 👍 

<img width="712" alt="Screenshot 2023-05-11 at 8 38 11 PM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/345be519-236a-4119-bc55-6612de16bba9">

**Comment by rflmota** [comment]
Hello @cfriedt , thank you for your tips, I tried them but wasn't successful when running on `qemu_cortex_m3` target.

![download-min](https://github.com/zephyrproject-rtos/zephyr/assets/5682257/a0b68be6-d59b-4f33-8284-76fec0a8c88a)

**Reply by cfriedt** [reply]
I don't believe `qemu_cortex_m3` is SMP capable. It's possible you will only see multiple threads when debugging an SMP-capable board such as `qemu_cortex_a53_smp`.

E.g. today, I was debugging several SMP Qemu targets and thread info was visible.

**Comment by noahp** [comment]
In this case, I think `qemu`'s gdbstub (server) would need to implement Zephyr rtos-awareness; i.e. in Jlink or OpenOCD, it's done by the server. I see some examples that do Zephyr rtos awareness on the client side, eg in CLion, if that helps!

**Comment by bdmendes** [comment]
I filed an issue for this, in case you are interested: https://github.com/zephyrproject-rtos/sdk-ng/issues/691

**Reply by rflmota** [reply]
Awesome work @bdmendes , thank you for sharing! 🙌🏻
